### PR TITLE
Cleanup the main anaconda.py file

### PR DESCRIPTION
**Pull request by M4rtinK**
General anaconda.py cleanup, refactoring and PEP8ization.

Functions and substantial pieces of code that don't really have to be in anaconda.py have been moved out to various modules, their names PEP8ized and docstrings added.

I'm sure things could be improved further:
- more code moved out or consolidated
- some of the long lists of functions being called could be ecapsulated to classes doing the same thing
- unit tests for the functions that have been moved/created
 -  most of them change system state, so it won't be easy or even doable in some cases
- there are still some UI parts that should be converted to signal that the UI connect to - no direct imports of UI stuff

But I guess this could do for now. :)

### CODE REVIEW

File: anaconda.py [hunk 1]
```diff
@@ -21,12 +21,14 @@
...
 # This toplevel file is a little messy at the moment... (2001-06-22)
 # ...still messy (2013-07-12)
+# A lot less messy now. :) (2016-10-13)
 
 import os
 import site
 
 coverage = None
 
+# setup code coeverage monitoring
```
Comment (dashea) [inline anaconda.py]: misspelled coverage

File: anaconda.py [hunk 2]
```diff
@@ -27,6 +27,7 @@
...
```
Comment (vpodzime) [inline anaconda.py]: *"coverage"*

File: anaconda.py [hunk 3]
```diff
@@ -21,6 +21,7 @@
 
 # This toplevel file is a little messy at the moment... (2001-06-22)
 # ...still messy (2013-07-12)
+# A lot less messy now. :) (2016-10-13)
```
Comment (vpodzime) [inline anaconda.py]: Nice!

**Comment by jkonecny12** [comment]
Please fix these Jenkins issues:

pylint:
```
************* Module ../pyanaconda/anaconda.py
W1201(logging-not-lazy):194,16: Anaconda.display_mode: Specify string format arguments as logging function parameters
W1201(logging-not-lazy):196,16: Anaconda.display_mode: Specify string format arguments as logging function parameters
```
Style guide:
```
Bad string HOSTNAME found at pyanaconda/startup_utils.py:294. Try host name instead.
Did not find 1 occurrences of HOSTNAME in anaconda.py
FAIL gettext_tests/style_guide.py (exit status: 1)
```

Storage tests:
```
Traceback (most recent call last):
  File "./storage/run_storage_tests.py", line 5, in <module>
    from pyanaconda.ui.common import collect
  File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/ui/__init__.py", line 23, in <module>
    from pyanaconda.ui.common import collect
  File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/ui/common.py", line 30, in <module>
    from pyanaconda import screen_access
  File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/screen_access.py", line 42, in <module>
    from pyanaconda import startup_utils
  File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/startup_utils.py", line 38, in <module>
    from pyanaconda import kickstart
  File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/kickstart.py", line 62, in <module>
    from pyanaconda.ui.common import collect
ImportError: cannot import name 'collect'
Exception ignored in: Exception ignored in: Exception ignored in: Exception ignored in: Exception ignored in: FAIL storage/run_storage_tests.py (exit status: 1)
```

File: anaconda.py [hunk 4]
```diff
@@ -739,9 +423,11 @@ def cleanPStore():
     from pyanaconda import kickstart
     from pyanaconda import ntp
     from pyanaconda import keyboard
+    from pyanaconda import display
+    from pyanaconda import startup_utils
```
Comment (rvykydal) [inline anaconda.py]: Not sure if it matters here, but startup_utils are imported about 40 lines above.

File: pyanaconda/startup_utils.py [hunk 5]
```diff
@@ -76,3 +87,101 @@ def get_anaconda_version_string():
     else:
         return "unknown"
 
+def gtk_warning(title, reason):
+    """A simple warning dialo for use during early startup of the Anaconda GUI.
```
Comment (rvykydal) [inline pyanaconda/startup_utils.py]: A typo, dialo -> dialog

**Comment by rvykydal** [comment]
Looks good to me. Really nice.

File: pyanaconda/startup_utils.py (hunk 5 above)
Comment (jkonecny12) [inline pyanaconda/startup_utils.py]: Or is this a typo dialo -> diablo ? :)

File: anaconda.py (hunk 1 above)
Comment (M4rtinK) [inline anaconda.py]: Fixed!

File: anaconda.py [hunk 6]
```diff
@@ -110,65 +110,6 @@ def exitHandler(rebootData, storage):
         else:  # reboot action is KS_REBOOT or None
             iutil.execWithRedirect("systemctl", ["--no-wall", "reboot"])
 
-def startSpiceVDAgent():
```
Comment (M4rtinK) [inline anaconda.py]: Fixed.

File: anaconda.py (hunk 2 above)
Comment (M4rtinK) [inline anaconda.py]: Fixed.

File: pyanaconda/display.py [hunk 7]
```diff
@@ -51,6 +51,11 @@
 # Spice
 
 def start_spice_vd_agent():
+    """Start the spice (vdagent).
+
+    For certain features to work spice requires that the guest os
+    is running the spice agent (vdagent).
+    """
```
Comment (M4rtinK) [inline pyanaconda/display.py]: Yeah, on a second look it looks better this way. Fixed. :)

File: pyanaconda/startup_utils.py (hunk 5 above)
Comment (M4rtinK) [inline pyanaconda/startup_utils.py]: Fixed. ;-)

**Comment by M4rtinK** [comment]
> Please fix these Jenkins issues:
> 
> pylint:
> 
> ************* Module ../pyanaconda/anaconda.py
> W1201(logging-not-lazy):194,16: Anaconda.display_mode: Specify string format arguments as logging function parameters
> W1201(logging-not-lazy):196,16: Anaconda.display_mode: Specify string format arguments as logging function parameters

Should be fixed now.

> 
> Style guide:
> 
> Bad string HOSTNAME found at pyanaconda/startup_utils.py:294. Try host name instead.
> Did not find 1 occurrences of HOSTNAME in anaconda.py
> FAIL gettext_tests/style_guide.py (exit status: 1)

It think this is actually a false positive - it points to:
`
stdout_log.info(_("Please ssh install@HOSTNAME to continue installation."))
`

This has been there before (which means the main anaconda.py file has probably never been scanned by the style guide) and HOSTNAME useage there is IMHO correct - a variable users should replace with concrete value, kinda like the meta-vars in argparse.

Can we somehow mark false positives for the style guide the way we can do it for Pylint ?

Alternatively we could change that string, but that seams kinda bacwards it would be basically working around a broken test. And it would change output on the s390.

> 
> Storage tests:
> 
> Traceback (most recent call last):
>   File "./storage/run_storage_tests.py", line 5, in <module>
>     from pyanaconda.ui.common import collect
>   File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/ui/__init__.py", line 23, in <module>
>     from pyanaconda.ui.common import collect
>   File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/ui/common.py", line 30, in <module>
>     from pyanaconda import screen_access
>   File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/screen_access.py", line 42, in <module>
>     from pyanaconda import startup_utils
>   File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/startup_utils.py", line 38, in <module>
>     from pyanaconda import kickstart
>   File "/home/jenkins/workspace/anaconda-master-PR/pyanaconda/kickstart.py", line 62, in <module>
>     from pyanaconda.ui.common import collect
> ImportError: cannot import name 'collect'
> Exception ignored in: Exception ignored in: Exception ignored in: Exception ignored in: Exception ignored in: FAIL storage/run_storage_tests.py (exit status: 1)

That traceback is pretty weird. Do I understand it correctly that the actual error is `ImportError: cannot import name 'collect'` ?

I haven't moved the collect() function - it still is in `pyanaconda/ui/common.py` as before. So maybe importing `pyanaconda/kickstart.py` from `startup_utils.py` (which is as far as I can tell the only relevant change introduced by this patch set) maybe somehow broke the import logic in `kickstart.py` ? It also all worked fine during normal GUI/TUI installation, so I guess it might also be influenced by different Python path configuration the unit tests have ?

**Comment by jkonecny12** [comment]
TL;DR
You can ignore pylint error now.

There is another pylint error but this error is caused by dnf incompatibility version. The problem here is that this PR is branched from DNF 1 version but there is DNF 2 version on the testing machines because rawhide is using DNF 2 now.

**Comment by jkonecny12** [comment]
The previous issue with importing ``collect`` looks fixed but you are getting new pylint errors:
```
************* Module ../pyanaconda/iutil.py
E0602(undefined-variable):1443,12: collect: Undefined variable 'imp'
E0602(undefined-variable):1444,46: collect: Undefined variable 'imp'
E0602(undefined-variable):1469,42: collect: Undefined variable 'types'
E0602(undefined-variable):1474,29: collect: Undefined variable 'imp'
E0602(undefined-variable):1505,15: collect: Undefined variable 'RemovedModuleError'
E0602(undefined-variable):1517,12: collect: Undefined variable 'imp'
E0602(undefined-variable):1522,24: collect.<lambda>: Undefined variable 'inspect'
E0602(undefined-variable):1526,22: collect: Undefined variable 'inspect'
************* Module ../pyanaconda/ui/common.py
E0602(undefined-variable):638,28: collect_spokes: Undefined variable 'collect'
E0602(undefined-variable):660,30: collect_categories: Undefined variable 'collect'
E0602(undefined-variable):663,30: collect_categories: Undefined variable 'collect'
W0611(unused-import): 20,0: : Unused import os
W0611(unused-import): 21,0: : Unused import imp
W0611(unused-import): 22,0: : Unused import inspect
W0611(unused-import): 23,0: : Unused import sys
W0611(unused-import): 24,0: : Unused import types
W0611(unused-import): 29,0: : Unused RemovedModuleError imported from pyanaconda.errors
```

File: anaconda.py [hunk 8]
```diff
@@ -1095,26 +591,30 @@ def _earlyExceptionHandler(ty, value, traceback):
...
         log.warning("reinitializing locale due to failed attempt to start the GUI")
-        localization.setup_locale(os.environ["LANG"], ksdata.lang, anaconda.displayMode != "g")
+        localization.setup_locale(os.environ["LANG"], ksdata.lang, text_mode=anaconda.tui_mode)
 
     # we now know in which mode we are going to run so store the information
-    from pykickstart.constants import DISPLAY_MODE_GRAPHICAL, DISPLAY_MODE_CMDLINE, DISPLAY_MODE_TEXT
-    mode_char_to_const = {'g': DISPLAY_MODE_GRAPHICAL, 't': DISPLAY_MODE_TEXT, 'c': DISPLAY_MODE_CMDLINE}
-    ksdata.displaymode.displayMode = mode_char_to_const[anaconda.displayMode]
+    from pykickstart import constants as pykickstart_constants
+    mode_char_to_const = {
```
Comment (jkonecny12) [inline anaconda.py]: There should be used some other name. The variable ``mode_char_to_const`` doesn't give much sense now.
What about ``display_mode_convertor``.

File: pyanaconda/display.py [hunk 9]
```diff
@@ -0,0 +1,328 @@
...
+        anaconda.display_mode = constants.DISPLAY_MODE_GUI
+        flags.usevnc = True
+        vnc_server.password = anaconda.ksdata.vnc.password
+
+def check_vnc_can_be_started(anaconda):
+    """Check if we can start VNC in the current environment.
+
+    :returns: if VNC can be started and list of possible reasons
+              why VNC can't be started
+    :rtype: bool and list
```
Comment (jkonecny12) [inline pyanaconda/display.py]: I would recommend to change the rtype to ``(boot, list)`` or ``tuple of (bool, list)``.

File: pyanaconda/display.py [hunk 10]
```diff
@@ -0,0 +1,328 @@
...
+    """
+
+    error_messages = []
+    vnc_startup_possible = True
+
+    # disable VNC over text question when not enough memory is available
+    if blivet.util.total_memory() < isys.MIN_GUI_RAM:
+        error_messages.append("Not asking for VNC because current memory (%d) < MIN_GUI_RAM (%d)" % (blivet.util.total_memory(), isys.MIN_GUI_RAM))
```
Comment (jkonecny12) [inline pyanaconda/display.py]: Please wrap this line.

File: pyanaconda/display.py [hunk 11]
```diff
@@ -0,0 +1,328 @@
...
+
+    if anaconda.gui_mode:
+        mods = (tup[1] for tup in pkgutil.iter_modules(pyanaconda.ui.__path__, "pyanaconda.ui."))
+        if "pyanaconda.ui.gui" not in mods:
+            stdout_log.warning("Graphical user interface not available, falling back to text mode")
+            anaconda.display_mode = constants.DISPLAY_MODE_TUI
+            flags.usevnc = False
+            flags.vncquestion = False
+
+    # check if VNC cna be started
```
Comment (jkonecny12) [inline pyanaconda/display.py]: typo ``cna`` -> ``can``

File: pyanaconda/display.py [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+            do_extra_x11_actions(options.runres, gui_mode=anaconda.gui_mode)
+
+    if anaconda.tui_mode and anaconda.gui_startup_failed and flags.vncquestion and not anaconda.ksdata.vnc.enabled:
+        message = _("X was unable to start on your "
+                    "machine.  Would you like to "
+                    "start VNC to connect to "
+                    "this computer from another "
+                    "computer and perform a "
+                    "graphical installation or continue "
+                    "with a text mode installation?")
```
Comment (jkonecny12) [inline pyanaconda/display.py]: Please merge these lines. I don't think there is an advantage to have so narrow lines here.

File: pyanaconda/startup_utils.py [hunk 13]
```diff
@@ -71,8 +92,365 @@ def get_anaconda_version_string():
         # Ignore pylint not finding the version module, since thanks to automake
         # there's a good chance that version.py is not in the same directory as
         # the rest of pyanaconda.
-        from pyanaconda import version # pylint: disable=no-name-in-module
+        from pyanaconda import version  # pylint: disable=no-name-in-module
```
Comment (jkonecny12) [inline pyanaconda/startup_utils.py]: Just a question, is this intentional? I looks to me like typo more than correction.

File: pyanaconda/startup_utils.py [hunk 14]
```diff
@@ -71,8 +92,365 @@ def get_anaconda_version_string():
...
+                      "install, but you only have %(total_ram)s MB on this machine.\n")
+    reason_graphical = _("The %(product_name)s graphical installer requires %(needed_ram)s "
+                         "MB of memory, but you only have %(total_ram)s MB\n.")
+
+    reboot_extra = _('\n'
+                     'Press [Enter] to reboot your system.\n')
+    livecd_title = _("Not enough RAM")
+    livecd_extra = _(" Try the text mode installer by running:\n\n"
+                     "'/usr/bin/liveinst -T'\n\n from a root "
+                     "terminal.")
```
Comment (jkonecny12) [inline pyanaconda/startup_utils.py]: Please move ``from a root terminal`` to one line

File: pyanaconda/startup_utils.py (hunk 13 above)
Comment (M4rtinK) [inline pyanaconda/startup_utils.py]: That's actually deliberate - [PEP8 specifies](https://www.python.org/dev/peps/pep-0008/#inline-comments) that incline comments should be prefixed by at least two spaces:

> An inline comment is a comment on the same line as a statement. Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.  

It's a minor thing but I'm trying to be consistent when at it. :)

**Comment by M4rtinK** [comment]
@jkonecny12 Thanks for noting! Fixed the latest bunch of Pylint issues that showed up after the collect() move patch was added.

File: pyanaconda/anaconda.py [hunk 15]
```diff
@@ -176,6 +176,43 @@ def storage(self):
...
+        return self._display_mode not in constants.NON_INTERACTIVE_DISPLAY_MODES
+
+    @display_mode.setter
+    def display_mode(self, new_mode):
+        new_mode_name = constants.DISPLAY_MODE_NAMES.get(new_mode)
+        if new_mode_name:
+            if self._display_mode:
+                old_mode = self._display_mode
+                old_mode_name = constants.DISPLAY_MODE_NAMES.get(old_mode)
+                log.debug("changing display mode from %s(%s) to %s(%s)", old_mode_name, old_mode, new_mode_name, new_mode)
```
Comment (M4rtinK) [inline pyanaconda/anaconda.py]: Wrapped it up.

File: anaconda.py (hunk 8 above)
Comment (M4rtinK) [inline anaconda.py]: Or `display_mode_conversion_table` ? I've used that for now.

File: pyanaconda/display.py (hunk 9 above)
Comment (M4rtinK) [inline pyanaconda/display.py]: OK.

File: pyanaconda/display.py (hunk 10 above)
Comment (M4rtinK) [inline pyanaconda/display.py]: Wrapped.

File: pyanaconda/display.py (hunk 11 above)
Comment (M4rtinK) [inline pyanaconda/display.py]: Fixed.

File: pyanaconda/display.py (hunk 12 above)
Comment (M4rtinK) [inline pyanaconda/display.py]: Yeah, it could be there for historic reasons, but it will be a single string anyway when the TUI widget gets it. So I've merged it to just 3 lines.

File: anaconda.py (hunk 8 above)
Comment (jkonecny12) [inline anaconda.py]: Sure, sounds great.

File: pyanaconda/startup_utils.py (hunk 13 above)
Comment (jkonecny12) [inline pyanaconda/startup_utils.py]: Ahh OK, thank you for explanation.

File: pyanaconda/startup_utils.py (hunk 14 above)
Comment (M4rtinK) [inline pyanaconda/startup_utils.py]: Fixed.
